include/ghex/unstructured/pattern.hpp

```diff
@@ -0,0 +1,340 @@
+                               << "partition = " << is.partition() << ";\n"
+                               << "local indexes: [ ";
+                            for (auto idx : is.local_index()) { os << idx << " "; }
+                            os << "]\n";
+                            return os;
+                        }
+
+                };
+
+                using iteration_space_pair = iteration_space;
+                using index_container_type = std::vector<iteration_space_pair>;
+
+                /** @brief extended domain id, including rank and tag information
+                 * WARN: domain id temporarily set equal to rank, differently from structured pattern case*/
+                struct extended_domain_id_type {
+
+                    // members
+                    domain_id_type id;
+                    int mpi_rank;
+                    address_type address;
+                    int tag;
+
+                    // member functions
+                    // /** @brief unique ordering given by id and tag*/
+                    // bool operator < (const extended_domain_id_type& other) const noexcept {
+                    //     return (id < other.id ? true : (id == other.id ? (tag < other.tag) : false));
+                    // }
+                    /** @brief unique ordering given by address and tag*/
+                    bool operator < (const extended_domain_id_type& other) const noexcept {
+                        return (address < other.address ? true : (address == other.address ? (tag < other.tag) : false));
+                    }
+
+                    /** @brief print*/
+                    template<class CharT, class Traits>
+                    friend std::basic_ostream<CharT, Traits>& operator << (std::basic_ostream<CharT, Traits>& os, const extended_domain_id_type& dom_id) {
+                        os << "{tag=" << dom_id.tag << ", rank=" << dom_id.mpi_rank << "}";
+                        return os;
+                    }
+
+                };
+
+                // halo map type
+                using map_type = std::map<extended_domain_id_type, index_container_type>;
+
+                // static member functions
+
+                /** @brief compute number of elements in an object of type index_container_type */
+                static int num_elements(const index_container_type& c) noexcept {
+                    std::size_t s{0};
+                    for (const auto& is : c) s += is.size();
+                    return s;
+                }
+
+            private:
+
+                // members
+                iteration_space_pair m_domain;
+                extended_domain_id_type m_id;
+                map_type m_send_map;
+                map_type m_recv_map;
+                pattern_container_type* m_container;
+
+            public:
+
+                // ctors
+                pattern(const iteration_space_pair& domain, const extended_domain_id_type& id) :
```

Comment:
should this be private and only accessible by the friend function make_pattern_impl?

Reply:
Probably yes, I used `public` basically to be consistent with structured pattern. Should we modify that one as well?